docs/guides/filesystem-usage.md

```diff
@@ -0,0 +1,191 @@
+
+## How to use lakeFS file system APIs
+
+The following section explains more in-depth how to use the `LakeFSFileSystem` APIs.
+This section concerns the explicitly implemented operations. In addition, there are a number of file system APIs inherited from the [`AbstractFileSystem` interface in fsspec](https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.spec.AbstractFileSystem).
+
+More information on file system usage can be found in the [`fsspec` documentation](https://filesystem-spec.readthedocs.io/en/latest/usage.html#use-a-file-system).
+
+### Uploading and downloading files
+
+The arguably most important feature of the file system is file transfers.
+
+#### File uploads
+
+To upload a file, you can use the `fs.put()` and `fs.put_file()` methods. 
+While `fs.put_file()` operates on single files only, the `fs.put()` API can be used for directory uploads.
+
+```python
+from lakefs_spec import LakeFSFileSystem
+
+fs = LakeFSFileSystem()
+
+# remote path, then local target path.
+fs.put_file("file.txt", "my-repo/my-ref/file.txt")
+```
+
+If you want to upload an entire directory to lakeFS, you can use the `fs.put()` API together with the `recursive=True` switch:
+
+```python
+# structure:
+#   dir/a.txt
+#      /b.yaml
+#      /c.csv
+#      /...
+
+fs.put("dir", "my-repo/my-ref/dir", recursive=True)
+```
+
+!!! info
+
+    The above method of file uploading results in two transfers: Once from the client to the lakeFS server, and once from the lakeFS server to the object storage.
+    This can impact performance if the uploaded files are very large. To avoid this performance hit, you can also decide to write the file directly to the underlying object storage:
+
+    ```python
+    fs = LakeFSFileSystem()
+    
+    fs.put_file("my-repo/my-ref/file.txt", "file.txt", use_blockstore=True)
+    ```
+
+    Direct lakeFS blockstore uploads require the installation of the corresponding `fsspec` file system implementation through `pip`.
+    For an S3-based lakeFS deployment, install the `s3fs` package. For Google Cloud Storage (GCS), install the `gcsfs` package.
+    For Azure blob storage, install the `adlfs` package.
+
+#### File downloads
+
+To download a file, you can use the `fs.get()` or `fs.get_file()` methods.
+While `fs.get_file()` downloads single files only, the `fs.get()` API can be used for recursive directory downloads.
+
+```python
+from lakefs_spec import LakeFSFileSystem
+
+fs = LakeFSFileSystem()
+
+# remote path, then local target path.
+fs.get_file("my-repo/my-ref/file.txt", "file.txt")
+```
+
+In the case of a directory in lakeFS, use the `fs.get()` API together with the `recursive=True` switch:
+
+```python
+# structure:
+#   dir/a.txt
+#      /b.yaml
+#      /c.csv
+#      /...
+
+# downloads the entire `dir` directory (and subdirectories) into the current directory.
+fs.get("my-repo/my-ref/dir", "dir", recursive=True)
```

Comment:
This does not work at all, but due to other reasons than for `put(..., recursive=True)`. It seems that the implementation of `AbstractFileSystem.expand_paths` might need to be overwritten for our case since it expands the paths without the `{repo}/{branch}/` prefix:

```
rpaths = self.expand_path(rpath, recursive=recursive, maxdepth=maxdepth)

# rpath = 'repo/main/data/'
# rpaths = ['data/weather-2010.json', 'data/weather-2010s.json', 'data/weather-2020.json', 'data/weather-2020s.json', 'repo/main/data/']
```

This leads to errors down the road when the code enters `get_file` for these rpaths, since they obviously don't `parse()` correctly.

Reply:
Addressed in 7e09b41c9f7d3ac7a0b2582832432358ebf4ff42 and through #166.